### PR TITLE
Index core classes using RBS

### PR DESCRIPTION
**Pull request by andyw8**
Step towards https://github.com/Shopify/ruby-lsp/issues/1335

To test:

- Disable the Sorbet extension and hover over a core class name, e.g. String.
- Verify you see the docs in the hover
- Use Go To Definition and verify it takes you to the RBS file with the String definitions.

After this we can move on to methods.

### CODE REVIEW

File: lib/ruby_lsp/server.rb
Comment (andyw8) [inline lib/ruby_lsp/server.rb:693]: Currently this is fast, so I don't think need to have any logging or progress indicator. Maybe we will once it's expanded to methods, etc.

**Comment by andyw8** [comment]
(seeing CI failures for older Rubies, looking into...)

**Review by st0012 — APPROVED** [review]
LGTM 👍 
Once this is merged, I have a branch that will introduce `Object`, `Kernel`, and `BasicObject` to the ancestors list.

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:5]: The name `IndexRbs` makes it sound like there are two indices: one for RBS and the other one.

Can we call this indexer, collector or something like that?

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:29]: Should we inline this inside the loop? Is there a need for the `process_signature` method?

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:43]: Since we're going to process multiple declaration types coming from RBS, can we already setup a case statement that only processes classes?

For example

```ruby
declarations.each do |declaration|
  case declaration
  when RBS::AST::Declarations::Class
    handle_class(declaration)
  else
    # Unhandled RBS declaration. Do nothing
  end
end
```

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:40]: Is this `nil` variable needed? Can we not get the parent class?

File: lib/ruby_indexer/test/configuration_test.rb
Comment (vinistock) [inline lib/ruby_indexer/test/configuration_test.rb:58]: I assume this is failing because now `abbrev` is a bundled dependency pulled by `rbs` rather than a default gem.

Just switch `abbrev` for some other default gem.

File: lib/ruby_lsp/server.rb
Comment (vinistock) [reply lib/ruby_lsp/server.rb:693]: How fast is it on a Core Spin instance?

**Review by vinistock — COMMENTED** [review]
This can be handled in the next PR, but we need to extract includes, prepends and extends and insert those as well.

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:5]: Let's use this occasion to rectify the naming, I think we should have one `Index`, the data structure, and a bunch of `XIndexer`, the things collecting data to populate the index: `RubyIndexer`, `RBSIndexer`, `RBIIndexer`, etc.

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:40]: You should be able to access it from the declaration: https://github.com/ruby/rbs/blob/64a96b134a9c2e4441d389f3895e3dcdba9f8938/lib/rbs/ast/declarations.rb#L92

File: lib/ruby_indexer/lib/ruby_indexer/location.rb
Comment (Morriar) [inline lib/ruby_indexer/lib/ruby_indexer/location.rb:14]: This may be better suited as a method of the RBS indexer to reduce coupling

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (andyw8) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:29]: Hmm, I prefer it this way for readability.

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (andyw8) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb:5]: I've renamed it to `RubyIndexer`.

@Morriar I agree with your suggestion, but we can do that in a follow-up?

File: ruby-lsp.gemspec
Comment (andyw8) [inline ruby-lsp.gemspec:23]: I verified it on 3.0.0. With 2.0.0 it fails because "each_signature" isn't defined.

Comment (andyw8) [reply ruby-lsp.gemspec:23]: (I'm open to removing the upper constraint)

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (andyw8) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb]: Thanks, updated.

File: lib/ruby_indexer/lib/ruby_indexer/rbs_indexer.rb
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/rbs_indexer.rb:19]: Do we need this method outside of this class? If not, it should be a private instance method.

File: ruby-lsp.gemspec
Comment (vinistock) [reply ruby-lsp.gemspec:23]: Let's keep the upper constraint. We can't guarantee that a major version won't break the LSP.

File: lib/ruby_lsp/server.rb
Comment (vinistock) [inline lib/ruby_lsp/server.rb:703]: We intentionally moved any sort of benchmarking to the client to avoid doing more work on the server. Let's not benchmark the execution time here please.

We can add benchmark scripts if necessary, but I honestly think it should be fine.

File: lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/index_rbs.rb]: Yep 👍

**Review by vinistock — APPROVED** [review]
Great work! 🚀

**Review by st0012 — APPROVED** [review]

**Comment by hsw15192617273** [comment]
hi, jruby maybe not be able to install rbs. 

```log
❯ gem install rbs                                                                                                                                                                                                                        (base)
Building native extensions. This could take a while...
ERROR:  Error installing rbs:
	ERROR: Failed to build gem native extension.

    current directory: /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/gems/shared/gems/rbs-3.5.1/ext/rbs_extension
/foo/bar/.rbenv/versions/jruby-9.4.7.0/bin/jruby -I /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib extconf.rb
checking for whether -std=gnu99 is accepted as CFLAGS... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/foo/bar/.rbenv/versions/jruby-9.4.7.0/bin/jruby
RuntimeError: The compiler failed to generate an executable file.
You have to install development tools first.

         try_do at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:456
    try_compile at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:571
    with_werror at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:522
    try_compile at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:571
     try_cflags at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:635
  append_cflags at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:641
   checking_for at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:942
       postpone at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:350
           open at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:320
       postpone at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:350
           open at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:320
       postpone at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:346
   checking_for at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:941
  append_cflags at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:640
           each at org/jruby/RubyArray.java:1981
  append_cflags at /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/stdlib/mkmf.rb:639
         <main> at extconf.rb:3

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/gems/shared/extensions/universal-java-17/3.1.0/rbs-3.5.1/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/gems/shared/gems/rbs-3.5.1 for inspection.
Results logged to /foo/bar/.rbenv/versions/jruby-9.4.7.0/lib/ruby/gems/shared/extensions/universal-java-17/3.1.0/rbs-3.5.1/gem_make.out
```

**Comment by andyw8** [comment]
@hsw15192617273 we don't currently support JRuby, but feel free to note this on https://github.com/Shopify/ruby-lsp/pull/1257

**Comment by vinistock** [comment]
@andyw8 sorry, I don't think that's accurate. The Ruby LSP does work on JRuby, we just didn't add it to CI.

Is it possible to add a gemspec constraint to only use `rbs` on non JRuby platforms?

**Comment by st0012** [comment]
> Is it possible to add a gemspec constraint to only use rbs on non JRuby platforms?

I don't think it's possible

> The Ruby LSP does work on JRuby, we just didn't add it to CI.

Wasn't it practically impossible because Sorbet can't run on it?

**Comment by andyw8** [comment]
@@hsw15192617273 was Ruby LSP working for you previously? (i.e. before RBS was introduced).

**Comment by hsw15192617273** [comment]
> @@hsw15192617273 was Ruby LSP working for you previously? (i.e. before RBS was introduced).

year, if install `ruby-lsp 0.16.7`, it's working fine. (I started using JRuby today)

**Comment by hsw15192617273** [comment]
but i not use sorbet in ruby code. just run ruby-lsp for vscode extension. JRuby install sorbet maybe not raise exception, I will confirm it tomorrow.

**Comment by soutaro** [comment]
I planned to implement a pure Ruby RBS parser to make it compatible to other Ruby implementations, but not finished yet...

**Comment by vinistock** [comment]
@andyw8 for the time being, to unblock our JRuby friends, let's remove the gemspec dependency from RBS and check for the presence of the gem instead.

```ruby
begin
  gem("rbs", ">= 3", "< 4")
rescue LoadError
  return
end

module RubyLsp
  class RBSIndexer
    # ...
```
And then where we use it

```ruby
index_rbs if defined?(RubyLsp::RBSIndexer)
```

**Pull request merged**
